### ISSUE TITLE
Gradle Plugin fails when importing protobufs with underscore in filename

### ISSUE THREAD

**Issue by dadadom** [salesforce/reactive-grpc#26]
I have the following protobuf definitions in `src/main/proto`:

`some_parameter.proto`:
```
syntax = "proto3";

package my.some_parameters;

message SomeParameter {
    string id = 1;
}
```
`my_service.proto`:
```
syntax = "proto3";

package my.some_service;

import "some_parameter.proto";

service MyService {
    rpc DoStuff (my.some_parameters.SomeParameter) returns (my.some_parameters.SomeParameter);
}
```

`gradle.build`:
```
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath "com.google.protobuf:protobuf-gradle-plugin:0.8.3"
    }
}

apply plugin: 'java'
apply plugin: 'com.google.protobuf'

repositories {
    mavenCentral()
}

dependencies {
    compile "com.google.protobuf:protobuf-java:3.4.0"
    compile "io.grpc:grpc-all:1.8.0"
    compile 'com.salesforce.servicelibs:reactor-grpc-stub:0.7.1'
}

protobuf {
    protoc {
        // Download from repositories
        artifact = "com.google.protobuf:protoc:3.4.0"
    }
    plugins {
        grpc {
            artifact = "io.grpc:protoc-gen-grpc-java:1.8.0"
        }
        reactor {
            artifact = "com.salesforce.servicelibs:reactor-grpc:0.7.1:jdk8@jar"
        }
    }
    generateProtoTasks {
        all().each { task ->
            task.plugins {
                grpc {}
                reactor {} // including this fails with:
                // /Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:28: error: package my.some_parameters.SomeParameter does not exist
                // public reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> doStuff(reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> reactorRequest) {
                //                                                                                                                                                        ^
                }
        }
    }
}
```

Gradle error message: 
```
gradle clean build
:clean
:extractIncludeProto
:extractProto UP-TO-DATE
:generateProto
:compileJava
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:28: error: package my.some_parameters.SomeParameter does not exist
        public reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> doStuff(reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> reactorRequest) {
                                                                                                                                                               ^
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:28: error: package my.some_parameters.SomeParameter does not exist
        public reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> doStuff(reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> reactorRequest) {
                                                                           ^
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:36: error: package my.some_parameters.SomeParameter does not exist
        public reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> doStuff(reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> request) {
                                                                                                                                                               ^
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:36: error: package my.some_parameters.SomeParameter does not exist
        public reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> doStuff(reactor.core.publisher.Mono<my.some_parameters.SomeParameter.SomeParameter> request) {
                                                                           ^
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:46: error: package my.some_parameters.SomeParameter does not exist
                                            my.some_parameters.SomeParameter.SomeParameter,
                                                                            ^
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:47: error: package my.some_parameters.SomeParameter does not exist
                                            my.some_parameters.SomeParameter.SomeParameter>(
                                                                            ^
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:73: error: package my.some_parameters.SomeParameter does not exist
                    com.salesforce.reactorgrpc.stub.ServerCalls.oneToOne((my.some_parameters.SomeParameter.SomeParameter) request,
                                                                                                          ^
/Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java:74: error: package my.some_parameters.SomeParameter does not exist
                            (io.grpc.stub.StreamObserver<my.some_parameters.SomeParameter.SomeParameter>) responseObserver,
                                                                                         ^
Note: /Users/dsandjaja/development/projects/testing/build/generated/source/proto/main/reactor/my/some_service/ReactorMyServiceGrpc.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
8 errors
:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 1.472 secs
```

If I rename `some_parameter.proto` to `someparameter.proto` and change the imports accordingly, compilation and Reactor class generation works.

It also works if I add the following to `some_parameter.proto`:
```
option java_outer_classname = "SomeParameterProto";
```

As it works without the reactor plugin, I would consider this a bug.

**Comment by rmichela** [comment]
Weird...I'll look into it.

**Comment by dadadom** [comment]
Works with the following ` pom.xml`:
``` 
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>mystuff</groupId>
    <artifactId>mytesting</artifactId>
    <version>0.0.1-SNAPSHOT</version>

    <properties>
        <maven.deploy.skip>true</maven.deploy.skip>
        <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>
    </properties>

    <dependencies>
        <dependency>
            <groupId>com.salesforce.servicelibs</groupId>
            <artifactId>reactor-grpc-stub</artifactId>
	    <version>0.7.1</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-core</artifactId>
	    <version>1.8.0</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-stub</artifactId>
	    <version>1.8.0</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-protobuf</artifactId>
	    <version>1.8.0</version>
        </dependency>
    </dependencies>

    <build>
        <extensions>
            <extension>
                <groupId>kr.motd.maven</groupId>
                <artifactId>os-maven-plugin</artifactId>
                <version>1.4.1.Final</version>
            </extension>
        </extensions>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-install-plugin</artifactId>
                <version>2.5.2</version>
                <configuration>
                    <skip>true</skip>
                </configuration>
            </plugin>

            <plugin>
                <groupId>org.xolstice.maven.plugins</groupId>
                <artifactId>protobuf-maven-plugin</artifactId>
                <version>0.5.1</version>
                <configuration>
                    <protocArtifact>com.google.protobuf:protoc:3.5.0:exe:${os.detected.classifier}</protocArtifact>
                    <pluginId>grpc-java</pluginId>
                    <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.8.0:exe:${os.detected.classifier}</pluginArtifact>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>test-compile</goal>
                            <goal>test-compile-custom</goal>
                        </goals>
                        <configuration>
                            <protocPlugins>
                                <protocPlugin>
                                    <id>reactor-grpc</id>
                                    <groupId>${project.groupId}</groupId>
                                    <artifactId>reactor-grpc</artifactId>
                                    <version>${project.version}</version>
                                    <mainClass>com.salesforce.reactorgrpc.ReactorGrpcGenerator</mainClass>
                                </protocPlugin>
                            </protocPlugins>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```

**Comment by rmichela** [comment]
@cbornet Any ideas?

**Comment by dadadom** [comment]
I just put the complete code in a new repository:
https://github.com/dadadom/failing-reactor-grpc/

**Comment by dadadom** [comment]
After some further investigation, it seems that this only fails iff the filename and the message name resolve to the same Java notation. In that case, the generated `inputType` and `outputType` for the mustache template, generated by the `ProtoTypeMap` from grpc-java-contrib seem to be wrongly generated.

I guess that the culprit is somewhere around https://github.com/salesforce/grpc-java-contrib/blob/master/jprotoc/src/main/java/com/salesforce/jprotoc/ProtoTypeMap.java#L54 , but I would have to investigate a bit further.

If needed, just close this issue and create a new one in https://github.com/salesforce/grpc-java-contrib

**Comment by rmichela** [comment]
## Root Cause

* Absent `option java_outer_classname`, protoc generates an outer java class for each .proto file with a name derived from the .proto file name.
* If no explicit java_outer_classname is specified in the .proto file, the class name will be constructed by converting the .proto file name to camel-case (so `foo_bar.proto` becomes `FooBar.java`)
* Protoc generates an inner class for each `message` type in the proto with a name that matches the `message` type name.
* In the event of a name conflict between the outer and inner type names, protoc adds an `OuterClass` suffix to the outer type's name.

As you have identified, JProtoc is not correctly accounting for the name collision between the outer and inner class name. 

## Workarounds

1. Add `option java_outer_classname = "SomeParameterProto";` to `some_parameter.proto`.
2. Rename `some_parameter.proto`.

## Mitigation
1. Fix JProtoc https://github.com/salesforce/grpc-java-contrib/issues/66
2. Update Reactive-gRPC generator

**Comment by rmichela** [comment]
https://github.com/salesforce/grpc-java-contrib/pull/67

**Comment by dadadom** [comment]
Any plans on releasing updated versions of jprotoc and reactive-grpc?

**Comment by rmichela** [comment]
Yes. Today

**Comment by rmichela** [comment]
https://github.com/salesforce/grpc-java-contrib/pull/70

**Comment by rmichela** [comment]
Released in 0.7.3

**Comment by dadadom** [comment]
You guys rock, thanks!
